launcher/FileIgnoreProxy.cpp

```diff
@@ -0,0 +1,256 @@
+ *      WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *      See the License for the specific language governing permissions and
+ *      limitations under the License.
+ */
+
+#include "FileIgnoreProxy.h"
+
+#include <QDebug>
+#include <QFileSystemModel>
+#include <QSortFilterProxyModel>
+#include <QStack>
+#include "FileSystem.h"
+#include "SeparatorPrefixTree.h"
+#include "StringUtils.h"
+
+FileIgnoreProxy::FileIgnoreProxy(QString root, QObject* parent) : QSortFilterProxyModel(parent), root(root) {}
+// NOTE: Sadly, we have to do sorting ourselves.
+bool FileIgnoreProxy::lessThan(const QModelIndex& left, const QModelIndex& right) const
+{
+    QFileSystemModel* fsm = qobject_cast<QFileSystemModel*>(sourceModel());
+    if (!fsm) {
+        return QSortFilterProxyModel::lessThan(left, right);
+    }
+    bool asc = sortOrder() == Qt::AscendingOrder ? true : false;
+
+    QFileInfo leftFileInfo = fsm->fileInfo(left);
+    QFileInfo rightFileInfo = fsm->fileInfo(right);
+
+    if (!leftFileInfo.isDir() && rightFileInfo.isDir()) {
+        return !asc;
+    }
+    if (leftFileInfo.isDir() && !rightFileInfo.isDir()) {
+        return asc;
+    }
+
+    // sort and proxy model breaks the original model...
+    if (sortColumn() == 0) {
+        return StringUtils::naturalCompare(leftFileInfo.fileName(), rightFileInfo.fileName(), Qt::CaseInsensitive) < 0;
+    }
+    if (sortColumn() == 1) {
+        auto leftSize = leftFileInfo.size();
+        auto rightSize = rightFileInfo.size();
+        if ((leftSize == rightSize) || (leftFileInfo.isDir() && rightFileInfo.isDir())) {
+            return StringUtils::naturalCompare(leftFileInfo.fileName(), rightFileInfo.fileName(), Qt::CaseInsensitive) < 0 ? asc : !asc;
+        }
+        return leftSize < rightSize;
+    }
+    return QSortFilterProxyModel::lessThan(left, right);
+}
+
+Qt::ItemFlags FileIgnoreProxy::flags(const QModelIndex& index) const
+{
+    if (!index.isValid())
+        return Qt::NoItemFlags;
+
+    auto sourceIndex = mapToSource(index);
+    Qt::ItemFlags flags = sourceIndex.flags();
+    if (index.column() == 0) {
+        flags |= Qt::ItemIsUserCheckable;
+        if (sourceIndex.model()->hasChildren(sourceIndex)) {
+            flags |= Qt::ItemIsAutoTristate;
+        }
+    }
+
+    return flags;
+}
+
+QVariant FileIgnoreProxy::data(const QModelIndex& index, int role) const
+{
+    QModelIndex sourceIndex = mapToSource(index);
+
+    if (index.column() == 0 && role == Qt::CheckStateRole) {
+        QFileSystemModel* fsm = qobject_cast<QFileSystemModel*>(sourceModel());
+        auto blockedPath = relPath(fsm->filePath(sourceIndex));
+        auto cover = blocked.cover(blockedPath);
+        if (!cover.isNull()) {
+            return QVariant(Qt::Unchecked);
+        } else if (blocked.exists(blockedPath)) {
+            return QVariant(Qt::PartiallyChecked);
+        } else {
+            return QVariant(Qt::Checked);
+        }
+    }
+
```

Comment:
Hmm, I guess it's performing a lot of unnecessary computation. I'll try changing that and you can let me know if it improves anything (or i could test)!

Reply:
Adding this... removes the icons which were there

Reply:
How about you limiting it to show icons at the top level?

Reply:
aside form the mentioned performance reasons you mean?
limiting it to show icons at the top level and preventing it form trying to request icons for every file in large folder like `mods` will greatly reduce the mentioned lag.

Reply:
yes

Reply:
I think it looks bad without icons tbh

Reply:
I suppose a checkbox in the dialogs (both the mrpack one and the zip export one) to toggle that would suffice. I'm not sure why the icons don't show up on my end, but I can attest that the GUI is too laggy for comfortable use in this scenario. Whether it's an issue in this code, on Qt, or on the OS / libs side of things, it doesn't matter.

Plus, I don't think it looks bad without icons :-)

Reply:
Now I look at it again it kind of looks better... I don't know what was going through my mind!
I don't think icons are the issue though. With this change it's a lot laggier than in 6.3

[Screencast from 2023-05-26 10-33-15.webm](https://github.com/PrismLauncher/PrismLauncher/assets/57493648/ed88054b-ca07-4b85-a526-15e0bd7b44fb)

the only thing I changed about the view component was https://github.com/PrismLauncher/PrismLauncher/pull/904#discussion_r1167548930 iirc (please ignore my messed up themes)

Reply:
For me, both the old implementation in 6.3 and your new implementation without the change I suggested are laggy, presumably due to the icons missing, while with those changes the laggyness (?) is gone.

Can you do a profiling to figure out what is slowing down on your end? I would suppose not having to look for icons would at least not worsen the performance of a list, so that's quite weird...

Reply:
Hmm, even without showing the icons it's still trying to load icons

Reply:
happens on the `develop` branch too...

Reply:
Yes, it's the same code...

> Hmm, even without showing the icons it's still trying to load icons

where are those loads coming from? It should be the case that returning early from this function would prevent Qt from going to look for icons.

Reply:
Ryex's suggestion seems better, probably best to discuss there. (:sob: this is getting messy)